base/broadcast.jl

```diff
@@ -3,63 +3,207 @@
-_containertype(::Type{<:Ptr}) = Any
-_containertype(::Type{<:Tuple}) = Tuple
-_containertype(::Type{<:Ref}) = Array
-_containertype(::Type{<:AbstractArray}) = Array
-_containertype(::Type{<:Nullable}) = Nullable
-containertype(x) = _containertype(typeof(x))
-containertype(ct1, ct2) = promote_containertype(containertype(ct1), containertype(ct2))
-@inline containertype(ct1, ct2, cts...) = promote_containertype(containertype(ct1), containertype(ct2, cts...))
-
-promote_containertype(::Type{Array}, ::Type{Array}) = Array
-promote_containertype(::Type{Array}, ct) = Array
-promote_containertype(ct, ::Type{Array}) = Array
-promote_containertype(::Type{Tuple}, ::ScalarType) = Tuple
-promote_containertype(::ScalarType, ::Type{Tuple}) = Tuple
-promote_containertype(::Type{Any}, ::Type{Nullable}) = Nullable
-promote_containertype(::Type{Nullable}, ::Type{Any}) = Nullable
-promote_containertype(::Type{T}, ::Type{T}) where {T} = T
-
-## Calculate the broadcast indices of the arguments, or error if incompatible
-# array inputs
-broadcast_indices() = ()
-broadcast_indices(A) = broadcast_indices(containertype(A), A)
-@inline broadcast_indices(A, B...) = broadcast_shape(broadcast_indices(A), broadcast_indices(B...))
-broadcast_indices(::ScalarType, A) = ()
-broadcast_indices(::Type{Tuple}, A) = (OneTo(length(A)),)
-broadcast_indices(::Type{Array}, A::Ref) = ()
-broadcast_indices(::Type{Array}, A) = indices(A)
+## logic for deciding the resulting container type
+# BottomArray dimensionality: computing max(M,N) in the type domain so we preserve inferrability
+_ruleba(::NTuple{N,Bool}) where N = BottomArray{N}
+longest(V1::Val, V2::Val) = longest(ntuple(identity, V1), ntuple(identity, V2))
+longest(t1::Tuple, t2::Tuple) = (true, longest(Base.tail(t1), Base.tail(t2))...)
+longest(::Tuple{}, t2::Tuple) = (true, longest((), Base.tail(t2))...)
+longest(t1::Tuple, ::Tuple{}) = (true, longest(Base.tail(t1), ())...)
+longest(::Tuple{}, ::Tuple{}) = ()
+
+# combine_types operates on values (arbitrarily many)
+combine_types(c) = result_type(rule(typeof(c)))
+combine_types(c1, c2) = result_type(combine_types(c1), combine_types(c2))
+combine_types(c1, c2, cs...) = result_type(combine_types(c1), combine_types(c2, cs...))
+
+# result_type works on types (singletons and pairs), and leverages `rule`
+result_type(::Type{T}) where T = T
+result_type(::Type{T}, ::Type{T}) where T     = T
+# Test both orders so users typically only have to declare one order
+result_type(::Type{S}, ::Type{T}) where {S,T} = result_join(S, T, rule(S, T), rule(T, S))
+
+# result_join is the final referee. Because `rule` for undeclared pairs results in Unknown,
+# we defer to any case where the result of `rule` is known.
+result_join(::Type{S}, ::Type{T}, ::Type{Unknown}, ::Type{Unknown}) where {S,T} = Unknown
+result_join(::Type{S}, ::Type{T}, ::Type{Unknown}, ::Type{U}) where {S,T,U}     = U
+result_join(::Type{S}, ::Type{T}, ::Type{U}, ::Type{Unknown}) where {S,T,U}     = U
```

Comment:
`result_join(::Type, ::Type, ::Type{U}, ::Type{Unknown}) where {U} = U` and similarly for the ones above?